dali/c_api_2/data_objects.h

```diff
@@ -0,0 +1,613 @@
+    auto &e = t_->ready_event();
+    if (e)
+      return e.get();
+    else
+      return std::nullopt;
+  }
+
+  cudaEvent_t GetOrCreateReadyEvent() override {
+    auto &e = t_->ready_event();
+    if (e)
+      return e.get();
+    int device_id = t_->device_id();
+    if (device_id < 0)
+      throw std::runtime_error("The tensor list is not associated with a CUDA device.");
+    t_->set_ready_event(CUDASharedEvent::Create(device_id));
+    return t_->ready_event().get();
+  }
+
+  daliTensorDesc_t GetDesc() const override {
+    auto &shape = t_->shape();
+    daliTensorDesc_t desc{};
+    desc.ndim = shape.sample_dim();
+    desc.data = t_->raw_mutable_data();
+    desc.dtype = t_->type();
+    desc.layout = GetLayout();
+    desc.shape = shape.data();
+    return desc;
+  }
+
+ private:
+  std::shared_ptr<Tensor<Backend>> t_;
+};
+
+template <typename Backend>
+RefCountedPtr<TensorWrapper<Backend>> Wrap(std::shared_ptr<Tensor<Backend>> tl) {
+  return RefCountedPtr<TensorWrapper<Backend>>(new TensorWrapper<Backend>(std::move(tl)));
+}
+
+template <typename Backend>
+class TensorListWrapper : public ITensorList {
+ public:
+  explicit TensorListWrapper(std::shared_ptr<TensorList<Backend>> tl) : tl_(std::move(tl)) {}
+
+  void Resize(
+      int num_samples,
+      int ndim,
+      const int64_t *shapes,
+      daliDataType_t dtype,
+      const char *layout) override {
+    Validate(dtype);
+    ValidateShape(num_samples, ndim, shapes);
+    if (layout)
+      Validate(TensorLayout(layout), ndim);
+    std::vector<int64_t> shape_data(shapes, shapes + ndim * num_samples);
+    tl_->Resize(TensorListShape<>(std::move(shape_data), num_samples, ndim), dtype);
+    if (layout)
+      tl_->SetLayout(layout);
+  }
+
+  void AttachBuffer(
+      int num_samples,
+      int ndim,
+      const int64_t *shapes,
+      daliDataType_t dtype,
+      const char *layout,
+      void *data,
+      const ptrdiff_t *sample_offsets,
+      daliDeleter_t deleter) override {
+    ValidateShape(num_samples, ndim, shapes);
+    Validate(dtype);
+
+    if (!data && num_samples > 0) {
+      for (int i = 0; i < num_samples; i++) {
+        auto sample_shape = make_cspan(&shapes[i*ndim], ndim);
+
+        if (volume(sample_shape) > 0)
+          throw std::invalid_argument(
+            "The pointer to the data buffer must not be null for a non-empty tensor list.");
+
+        if (sample_offsets && sample_offsets[i])
+          throw std::invalid_argument(
+            "All sample_offsets must be zero when the data pointer is NULL.");
+      }
+    }
+
+    TensorLayout new_layout = {};
+
+    if (!layout) {
+      if (ndim == tl_->sample_dim())
+        new_layout = tl_->GetLayout();
+    } else {
+      new_layout = layout;
+      Validate(new_layout, ndim);
+    }
+
+    tl_->Reset();
+    tl_->SetSize(num_samples);
+    tl_->set_sample_dim(ndim);
+    tl_->SetLayout(new_layout);
+    tl_->set_type(dtype);
```

Comment:
Why sometimes it's `like_that` and sometimes `LikeThat`?

Reply:
It's a mess - but It's also beyond the scope of this PR, isn't it?